create_all_tags.py

```diff
@@ -0,0 +1,129 @@
+REQUIRED_TAGS = [
+    "v2.0.1",
+    "v2.0.4",
+    "v2.1.0",
+    "v2.10.1",
+    "v2.12.0",
+    "v2.13.0",
+    "v2.14.0",
+    "v2.16.0",
+    "v2.17.0",
+    "v2.19.0",
+    "v2.2.0",
+    "v2.20.0",
+    "v2.21.0",
+    "v2.22.0",
+    "v2.23.0",
+    "v2.24.0",
+    "v2.24.1",
+    "v2.25.0",
+    "v2.26.1",
+    "v2.27.0",
+    "v2.28.0",
+    "v2.29.0",
+    "v2.3.0",
+    "v2.30.0",
+    "v2.31.0",
+    "v2.32.0",
+    "v2.33.0",
+    "v2.34.0",
+    "v2.4.0",
+    "v2.4.1",
+    "v2.4.2",
+    "v2.4.3",
+    "v2.4.4",
+    "v2.5.0",
+    "v2.6.1",
+    "v2.7.0",
+    "v2.7.1",
+    "v2.8.0",
+    "v2.9.0",
+]
+
+if __name__ == "__main__":
+    errors = []
+    for tag in REQUIRED_TAGS:
+        try:
+            fetched = repo.git.fetch(
+                # adding --no-tags to prevent all tags from being added
+                f"origin --no-tags refs/tags/{tag}:refs/tags/{tag}".split()
+            )
+            print(f"Already exists at origin: {tag}")
+            continue
+        except Exception:
+            pass
+        repo.git.fetch(f"upstream --no-tags refs/tags/{tag}:refs/tags/{tag}".split())
+        print(f"Fetched from upstream: {tag}")
+        # Delete local tag to allow for checkout
+        repo.git.tag(f"-d {tag}".split())
+        repo.git.checkout(f"tags/{tag}")
+
+        file_path = ""
+        for file in KNOWN_LOCATIONS_OF_STATE_GO:
+            if os.path.isfile(file):
+                print(f"found file at {file}")
+                file_path = file
+
+        if not file_path:
+            print(f"file not found, skipping {tag}")
+            errors.append(f"Tag: {tag}, File not found")
+            continue
+
+        with open(file_path, "r") as f:
+            content = f.readlines()
+
+        patched = False
+        for i, line in enumerate(content):
+            if line in KNOWN_FUNC_DECLARATION:
+                content.insert(i + 1, "\n".join(PATCH))
+                patched = True
+                break
+
+        if not patched:
+            print(f"Did not find function declaration in {file_path} for {tag}")
+            errors.append(f"Tag: {tag} Function declaration not found.")
+            continue
+
+        with open(file_path, "w") as f:
+            f.writelines(content)
+
+        print(repo.git.diff("**/state.go"))
+        user_input = input(
+            f"\n\nModification to {file_path}\nIf approved press <enter>. Any other value will cancel the operation.\n"
+            "If <enter> is pressed the changes to this file will be commited and a tag will be created and pushed."
+        )
+        if user_input != "":
+            repo.git.checkout(f"tags/{tag} {file_path}".split())
+            continue
+
+        repo.git.add(file_path)
+        repo.git.commit(f"-m 'creating tag for version {tag}'")
+        repo.git.tag(f"-d {tag}".split())
+        repo.git.tag(tag)
+        repo.git.push(f"origin tag {tag}".split())
```

Comment:
Wondering if `git push origin --tags` after the for-loop would work better. That way, each patch isn't delayed by pushing to remote.

Reply:
Good idea, I have a couple more enhancement to the way I am using gitpython to push as well. I will add that to the list of improvements